pkg/kor/hpas.go

```diff
@@ -0,0 +1,151 @@
+package kor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+	"k8s.io/utils/strings/slices"
+	"log"
+	"os"
+)
+
+type Hpa struct {
+	TargetKind string
+	TargetName string
+	Name       string
+	Namespace  string
+}
+
+func getDeploymentNames(clientset *kubernetes.Clientset, namespace string) ([]string, error) {
+	deployments, err := clientset.AppsV1().Deployments(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, 0, len(deployments.Items))
+	for _, deployment := range deployments.Items {
+		names = append(names, deployment.Name)
+	}
+	return names, nil
+}
+
+func getStatefulSetNames(clientset *kubernetes.Clientset, namespace string) ([]string, error) {
+	statefulsets, err := clientset.AppsV1().StatefulSets(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	names := make([]string, 0, len(statefulsets.Items))
+	for _, statefulset := range statefulsets.Items {
+		names = append(names, statefulset.Name)
+	}
+	return names, nil
+}
+
+func getHpas(clientset *kubernetes.Clientset, namespace string) ([]Hpa, error) {
+	rawHpas, err := clientset.AutoscalingV1().HorizontalPodAutoscalers(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	hpas := make([]Hpa, 0, len(rawHpas.Items))
+	for _, hpa := range rawHpas.Items {
+		hpas = append(hpas, Hpa{
+			TargetKind: hpa.Spec.ScaleTargetRef.Kind,
+			TargetName: hpa.Spec.ScaleTargetRef.Name,
+			Name:       hpa.Name,
+			Namespace:  hpa.Namespace,
+		})
+	}
+	return hpas, nil
+}
+
+func extractUnusedHpas(clientset *kubernetes.Clientset, namespace string) ([]string, error) {
+	deploymentNames, err := getDeploymentNames(clientset, namespace)
+	if err != nil {
+		return nil, err
+	}
+	statefulsetNames, err := getStatefulSetNames(clientset, namespace)
+	if err != nil {
+		return nil, err
+	}
+	hpas, err := getHpas(clientset, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	var diff []string
+	for _, hpa := range hpas {
+		if hpa.TargetKind == "Deployment" {
+			if !slices.Contains(deploymentNames, hpa.TargetName) {
+				diff = append(diff, hpa.Name)
+			}
+		} else if hpa.TargetKind == "StatefulSet" {
+			if !slices.Contains(statefulsetNames, hpa.TargetName) {
+				diff = append(diff, hpa.Name)
+			}
+		}
+	}
+	return diff, nil
+}
+
+func ProcessNamespaceHpas(clientset *kubernetes.Clientset, namespace string) ([]string, error) {
```

Comment:
does this method need to be public? I don't see it being used outside of this package.